firedrake/plot.py

```diff
@@ -292,6 +294,321 @@ def quiver(function, **kwargs):
     return axes.quiver(*(coords.T), *(vals.T), C, **kwargs)
 
 
+def _step_to_boundary(function, x, u, dt, loc_tolerance):
+    bracket = (0., dt)
+    while bracket[1] - bracket[0] > loc_tolerance * dt:
+        ds = (bracket[1] + bracket[0]) / 2
+        try:
+            y = x + ds * u
+            function.at(y, tolerance=loc_tolerance)
+            bracket = (ds, bracket[1])
+        except PointNotInDomainError:
+            bracket = (bracket[0], ds)
+
+    return bracket[0]
+
+
+def streamline(function, point, direction=+1, tolerance=3e-3, loc_tolerance=1e-10):
+    r"""Generate a streamline of a vector field starting from a point
+
+    :arg function: the Firedrake :class:`~.Function` to plot
+    :arg point: the starting point of the streamline
+    :arg direction: either +1 or -1 to integrate forward or backward
+    :arg tolerance: dimensionless tolerance for the RK12 adaptive integration
+    :arg loc_tolerance: tolerance for point location
+    :yield x, v, dt: the position, velocity, and timestep
+    """
+    mesh = function.ufl_domain()
+    cell_sizes = mesh.cell_sizes
+
+    x = np.array(point)
+    v1 = direction * function.at(x, tolerance=loc_tolerance)
+    r = cell_sizes.at(x, tolerance=loc_tolerance)
```

Comment:
All that you really need is some length scale to guide the adaptivity, so I don't think it matters one way or the other for the numerics.

At one point I did write this using a `cell_sizes` object by interpolating into DG0, but in order to avoid recreating the cell sizes for every streamline, I had to wrap the generator in a whole other class and the code was more complicated. If you think it's warranted I can go back to doing things that way.

Reply:
Ah, ok, that's fine.